### PR TITLE
fix: ANRs when getting identities [WPB-8753]

### PR DESCRIPTION
**Pull request by saleniuk**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

ANR when trying to get user or device identities.

### Solutions

`ObserveConversationMembersUseCase` was emitting the same items multiple times, especially when navigating in the app, and it was causing the spam of `getUserIdentities` and probably overloading `CoreCrypto`, so `distinctUntilChanged` worked.

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

Navigate: group conversation -> details -> participants -> user profile -> 1:1 conversation -> back or conversation details

### Attachments (Optional)

ANR_getclientidentity

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt [hunk 1]
```diff
@@ -895,3 +903,5 @@ internal class MLSConversationDataSource(
 
     private data class CommitOperationResult<T>(val commitBundle: CommitBundle?, val result: T)
 }
+
+val IDENTITIES_EXPIRATION = 1.seconds
```
Comment (yamilmedina) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt:907]:
```suggestion
val IDENTITIES_TTL = 1.seconds
```

Just a suggestion to have this named as the common patter for cache, `time to live`.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/util/ExpirableCache.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.util
+
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.datetime.Instant
+import kotlin.time.Duration
+
+class ExpirableCache<K, V>(private val expiration: Duration, private val currentTime: CurrentTimeProvider) {
```
Comment (yamilmedina) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/util/ExpirableCache.kt:25]:
```suggestion
class ExpirableCache<K, V>(private val timeToLive: Duration, private val currentTime: CurrentTimeProvider) {
```

Just a suggestion to have this named as the common patter for cache, `time to live`.

**Review by yamilmedina — APPROVED** [review]
🚀  Just small naming suggestion, cool idea about the `CurrentTimeProvider`

**Review by MohamadJaara — COMMENTED** [review]
tbh, i am not sure about the cache for user identities, i would rather fix the root cause, `ObserveParticipantsForConversationUseCase`. is calling this function on each update, maybe we can change this  to not fetch this info on each flow emit but have it fetched separately in the view model when needed?

**Review by MohamadJaara — APPROVED** [review]
great work 👍🏼

**Pull request merged**